polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp

```diff
@@ -924,13 +1144,32 @@ static size_t moveLoopInvariantCode(LoopLikeOpInterface loop,
 
   LoopGuardBuilder::create(loop)->guardLoop();
 
-  size_t numOpsHoisted = 0;
+  size_t numVersion = 0;
+  std::set<const Operation *> opsHoisted;
   for (const LICMCandidate &candidate : LICMCandidates) {
-    loop.moveOutOfLoop(&candidate.getOperation());
-    ++numOpsHoisted;
+    // Cannot hoist if any of its prerequisites are not hoisted.
+    if (any_of(candidate.getPrerequisites(),
+               [&opsHoisted](Operation *prerequisite) {
+                 return !opsHoisted.count(prerequisite);
+               }))
+      continue;
+
+    AccessorPairsType AccessorPairs =
+        candidate.getRequireNoOverlapAccessorPairs();
+    // Currently only version for single accessor pair.
+    if (AccessorPairs.size() > 1)
+      continue;
+    if (AccessorPairs.size() == 1 && numVersion < VersionLimit) {
+      SYCLAccessorVersionBuilder(loop, AccessorPairs[0]).versionLoop();
+      ++numVersion;
+    }
+
+    Operation *op = &candidate.getOperation();
+    loop.moveOutOfLoop(op);
+    opsHoisted.insert(op);
   }
 
-  return numOpsHoisted;
+  return opsHoisted.size();
```

Comment:
The current function `moveLoopInvariantCode` should (after collecting candidates) just concern itself with the mechanics of hoisting the candidates (that might require versioning the loop).  This code should be placed in some "pruning/filtering" function that decide which candidates to remove based on a set of heuristics (e.g. max number of versioning condition equal to N).